sal_interfaces/PMD/PMD_Telemetry.xml

```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?xml-stylesheet type="text/xsl" href="http://lsst-sal.tuc.noao.edu/schema/SALTelemetrySet.xsl"?>
+<SALTelemetrySet xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://lsst-sal.tuc.noao.edu/schema/SALTelemetrySet.xsd">
+  <SALTelemetry>
+    <Subsystem>PMD</Subsystem>
+    <EFDB_Topic>PMD_position</EFDB_Topic>
+    <item>
+      <EFDB_Name>position</EFDB_Name>
```

Comment:
The `position` telemetry topic is a single measurement, with no identifying information. That suggest CSC will only be capable of reading a single micrometer, and that each micrometer will have its own SAL index. That is only acceptable if you are sure that each micrometer can be read completely independently of all others (e.g. one TCP/IP or USB or RS-232 interface per micrometer). In the Mitutoyo systems I used to use we had a single box that read multiple micrometers. Your design doesn't support that.

It also means constructing many domain participants, each one of which only reads a single micrometer. That is wasteful of resources (though using the shared memory model probably makes it tolerable), and keeping track of the state of many different CSCs. In other words: it doesn't scale very well.

I suspect the simplest solution in ts_xml is to add a "name" or "label" string field to this topic, and probably a timestamp (double) field as well. That way your CSC can theoretically (at least) report data for more than one micrometer. It can even report the exact time a measurement was made even if you read N micrometers at once (though if the exact time isn't crucial then it suffices to use the private_sndStamp field -- the time at which the message was sent; it just means that if you read 4 micrometers at once the times will be slightly off for 3 of them).

My big fear is that will make the data in the EFD difficult to "mine". I am assuming it is practical to filter the values by "name", but I don't know that for a fact.

Reply:
You're right of course. The hub I have now reads from up to 8 devices. So should I make the count 8 for the position item or is there a more appropriate alternative?

Reply:
I think adding an additional "name" or "label" field would probably be more suitable. So each topic published can be associated with a "device or port".  It does make mining the EFD a bit trickier but still feasible. Both chronograph and DM's python library support filtering using a label. 

Having a vector of 8 values is probably not the best, since it restricts to the particular hardware you have available now.

Reply:
If you agree to make the CSC read multiple boxes then my suggestion is to make this a long array (e.g. on the order of 100 elements) coupled with:
* Set unused values to NaN so the EFD does not ingest them.
* Publish a list of micrometer labels as an event (not telemetry, since it is constant). These labels tell the user what each index is measuring, and which indices have no associated micrometer (blank label).
* The list of labels will have to be a single string, instead of an array of strings (SAL does not support arrays of strings).
* Use commas to separate the names and ignore whitespace after each comma. (I wish Python had direct support for splitting a string this way, but it can be done with re or by splitting on comma and then stripping each resulting name).
* If the list contains a blank label then you know that micrometer does not exist. This is important because you might not have all inputs in a given micrometer interface populated.
* Similarly I suggest making the list of labels only as long as needed, rather than requiring the configuration to list 100 names, most of which are blank.

If the PMD reads only one box then the above is still a reasonable way to go, but with a shorter list. How short? Do you assume 8 is the limit because your current interface box can only read 8 micrometers? Do you try to avoid future schema changes (e.g. in case we upgrade to a future box that can read more micrometers) by making the list longer? Probably 8 is fine, but XML schema changes are disruptive.

Reply:
Before we proceed with this discussion. 

DDS seems to support strings with unrestricted number of characters. Isn't that the case for arrays as well? It would be much better if we could dynamically adjust the size of the array instead of having an array of 1000 elements and we have to fiddle around with `nan` and all that.

Reply:
Do we need to have a meeting about this and other issues?

Reply:
@tribeiro SAL only supports fixed-length arrays (and does not support arrays of strings at all). I realize DDS is more flexible. Dave had some reason for imposing these restrictions. I don't know how difficult it would be to expand and I worry a bit what would the EFD entries would look like. I personally suggest living with the current SAL limitations for this CSC.

Reply:
@tribeiro @patrickingraham @couger01 we talked today and I think we agreed on the following (please give feedback for anything I got wrong):
* Use one SAL index per measurement aggregator device (e.g. Mitutoyo 8-micrometer box). That was the plan all along, we just confirmed it.
* Output measurements as an array of 8 values for now. Tiago will talk to Dave about adding support for variable-length arrays to SAL, but that will do for now.
* Output NaN for unused slots (and for readings that fail) in that 8 element array.
* Add a "metadata" event to describe the measurements. It should have the following fields (at least). Note that  I made some changes on as I wrote it up and thought about it, so definitely push back if you disagree:

  * kind: a single string describing the kind of measurement, e.g. "micrometers" or "incremental rotary encoders"
  * location: a string that gives the overall location of the measurement devices, e.g. "MTM1M3 Mirror Cell"
  * names: a string that is a of comma-separated names that describe in detail what each measurement device measures. Provide an explicit empty entry for each unused slot. Ignore whitespace after a comma. For example if location="MTM1M3 Mirror Cell" then names might be "x, y, zA, zB, zC, , , "
  * units: a string that gives the units of measurement, e.g. "um". I assume all devices output the same units so we just need one entry. If that is not a safe assumption please push back and we can make the string a comma-separated list of units. That will affect the description string for the field, but not the name or type of the field.